spec/referee_spec.rb

```diff
@@ -3,106 +3,110 @@
 
 describe "Referee" do
   let(:referee) { Referee.new }
+  let(:board) { Board.new }
 
   it "returns horizontal winner" do
-    board = Board.new
-    board.put_token_in_board([1, -1], 'X')
-    board.put_token_in_board([2, -1], 'X')
-    board.put_token_in_board([3, -1], 'X')
+    board.set_token_at(1, 'X')
+    board.set_token_at(2, 'X')
+    board.set_token_at(3, 'X')
+
     expect(referee.winner?(board)).to eq(true)
   end
 
   it "returns vertical winner" do
-    board1 = Board.new
-    board1.put_token_in_board([1, -1], 'X')
-    board1.put_token_in_board([4, -1], 'X')
-    board1.put_token_in_board([7, -1], 'X')
-    expect(referee.winner?(board1)).to eq(true)
+    board.set_token_at(1, 'X')
+    board.set_token_at(4, 'X')
+    board.set_token_at(7, 'X')
+
+    expect(referee.winner?(board)).to eq(true)
   end
 
   it "returns diagonal winner" do
-    board2 = Board.new
-    board2.put_token_in_board([1, -1], 'X')
-    board2.put_token_in_board([5, -1], 'X')
-    board2.put_token_in_board([9, -1], 'X')
-    expect(referee.winner?(board2)).to eq(true)
+    board.set_token_at(1, 'X')
+    board.set_token_at(5, 'X')
+    board.set_token_at(9, 'X')
+
+    expect(referee.winner?(board)).to eq(true)
   end
 
   it "returns true if tie exists" do
-    board = Board.new
     board.tokens = {
                    0 => 'O', 1 => 'X', 2 => 'X',
                    3 => 'X', 4 => 'X', 5 => 'O',
                    6 => 'O', 7 => 'O', 8 => 'X'
                   }
+
     expect(referee.tie?(board)).to eq(true)
   end
 
   describe "game_over?" do
     it "returns true if board is full & there's a tie" do
-      board = Board.new
       board.tokens = {
                      0 => 'O', 1 => 'X', 2 => 'X',
                      3 => 'X', 4 => 'X', 5 => 'O',
                      6 => 'O', 7 => 'O', 8 => '9'
                     }
+
       expect(referee.game_over?(board)).to eq(false)
 
-      board.put_token_in_board([9, -1], 'X')
+      board.set_token_at(9, 'X')
+
       expect(referee.game_over?(board)).to eq(true)
       expect(referee.winner?(board)).to eq(false)
     end
 
     it "returns true if there's a winner" do
-      board = Board.new
       board.tokens = {
                      0 => '1', 1 => 'X', 2 => 'X',
                      3 => 'X', 4 => 'X', 5 => 'O',
                      6 => 'O', 7 => 'O', 8 => '9'
                     }
```

Comment:
evita usar los internals ("tokens") – en otras palabras, evita leaky abstractions 
- https://www.joelonsoftware.com/2002/11/11/the-law-of-leaky-abstractions/
- https://en.wikipedia.org/wiki/Leaky_abstraction

por otro lado, me agrada tokens = { ... } hace que el tablero sea casi visible; ayuda en estas pruebas de referee entender pq un setup será ganador o no

```rb
# mucho mas dificil de visualizar que ganas en 9, X
set_token_at(1,X)
set_token_at(2,O)
set_token_at(5,X)
set_token_at(6,O)
set_token_at(7,X)
```

después de leer los artículos, evita la leaky abstraction, y para mantener la visibilidad te recomiendo usar un método helper en tus pruebas que acepte un input visible / claro, y marque el board

ej:

```rb
board = test_board(%w(
  _ X X
  X X O
  O O _
))

board.set_token_at(1, 'X')

expect(referee.game_over?(board)).to eq(true)
expect(referee.winner?(board)).to eq(true)
```

esto lo cubren en Clean Code, creo que el ejemplo eran unos tests de temperatura. Menciona usar un "test DSL" (domain specific lang)

Reply:
Done